Description: Force enable 16K alignment on Android. Must only be applied to 64
bit builds. Based on a patch to Krita by Dmitry Kazakov.

--- a/mkspecs/android-clang/qmake.conf
+++ b/mkspecs/android-clang/qmake.conf
@@ -68,8 +68,8 @@
 QMAKE_CFLAGS_HIDESYMS   = -fvisibility=hidden
 QMAKE_CFLAGS_NEON       = -mfpu=neon
 
-QMAKE_LFLAGS_APP        = -Wl,--build-id=none -Wl,--no-undefined -Wl,-z,noexecstack -shared
-QMAKE_LFLAGS_SHLIB      = -Wl,--build-id=none -Wl,--no-undefined -Wl,-z,noexecstack -shared
+QMAKE_LFLAGS_APP        = -Wl,--build-id=none -Wl,--no-undefined -Wl,-z,noexecstack -shared -Wl,-z,max-page-size=16384 -Wl,-z,common-page-size=16384
+QMAKE_LFLAGS_SHLIB      = -Wl,--build-id=none -Wl,--no-undefined -Wl,-z,noexecstack -shared -Wl,-z,max-page-size=16384 -Wl,-z,common-page-size=16384
 QMAKE_LFLAGS_PLUGIN     = $$QMAKE_LFLAGS_SHLIB
 QMAKE_LFLAGS_NOUNDEF    = -Wl,--no-undefined
 QMAKE_LFLAGS_RPATH      = -Wl,-rpath=

--- a/src/android/templates/build.gradle
+++ b/src/android/templates/build.gradle
@@ -41,7 +41,11 @@ android {
     ndkVersion androidNdkVersion
 
     // Extract native libraries from the APK
-    packagingOptions.jniLibs.useLegacyPackaging true
+    packagingOptions {
+        jniLibs {
+            useLegacyPackaging = true
+        }
+    }
 
     sourceSets {
         main {
@@ -71,6 +75,7 @@ android {
 
     // Do not compress Qt binary resources file
     aaptOptions {
+        noCompress 'so'
         noCompress 'rcc'
     }
 
--- a/src/tools/androiddeployqt/main.cpp
+++ b/src/tools/androiddeployqt/main.cpp
@@ -2785,7 +2785,7 @@ bool signPackage(const Options &options)
     };
 
     const QString verifyZipAlignCommandLine =
-            QLatin1String("%1%2 -c 4 %3")
+            QLatin1String("%1%2 -c -P 16 4 %3")
                     .arg(shellQuote(zipAlignTool),
                          options.verbose ? QLatin1String(" -v") : QLatin1String(),
                          shellQuote(packagePath(options, UnsignedAPK)));
@@ -2806,7 +2806,7 @@ bool signPackage(const Options &options)
             fprintf(stdout, "APK not aligned, aligning it for signing.\n");
 
         const QString zipAlignCommandLine =
-                QLatin1String("%1%2 -f 4 %3 %4")
+                QLatin1String("%1%2 -f -P 16 4 %3 %4")
                         .arg(shellQuote(zipAlignTool),
                              options.verbose ? QLatin1String(" -v") : QLatin1String(),
                              shellQuote(packagePath(options, UnsignedAPK)),
